backend/app/auth/handlers.py

```diff
@@ -0,0 +1,120 @@
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Annotated
+
+import jwt
+import requests
+from fastapi import Depends, HTTPException, Request
+from fastapi.security.http import HTTPBearer
+
+import app.storage as storage
+from app.auth.settings import AuthType, settings
+from app.schema import User
+
+
+class AuthHandler(ABC):
+    @abstractmethod
+    async def __call__(self, request: Request) -> User:
+        """Auth handler that returns a user object or raises an HTTPException."""
+
+
+class NOOPAuth(AuthHandler):
+    _default_sub = "static-default-user-id"
+
+    async def __call__(self, request: Request) -> User:
+        sub = request.cookies.get("opengpts_user_id") or self._default_sub
+        user, _ = await storage.get_or_create_user(sub)
+        return user
+
+
+class JWTAuthBase(AuthHandler):
+    async def __call__(self, request: Request) -> User:
+        http_bearer = await HTTPBearer()(request)
+        token = http_bearer.credentials
+
+        try:
+            payload = self.decode_token(token, self.get_decode_key(token))
+        except jwt.PyJWTError as e:
+            raise HTTPException(status_code=401, detail=str(e))
+
+        user, _ = await storage.get_or_create_user(payload["sub"])
+        return user
+
+    @abstractmethod
+    def decode_token(self, token: str, decode_key: str) -> dict:
+        ...
+
+    @abstractmethod
+    def get_decode_key(self, token: str) -> str:
+        ...
+
+
+class JWTAuthBasic(JWTAuthBase):
+    """Auth handler that uses a hardcoded decode key from env."""
+
+    def decode_token(self, token: str, decode_key: str) -> dict:
+        return jwt.decode(
+            token,
+            decode_key,
+            issuer=settings.jwt_basic.iss,
+            audience=settings.jwt_basic.aud,
+            algorithms=["HS256", "RS256"],
+            options={"require": ["exp", "iss", "aud", "sub"]},
+        )
+
+    def get_decode_key(self, token: str) -> str:
+        return settings.jwt_basic.decode_key
+
+
+class JWTAuthOIDC(JWTAuthBase):
+    """Auth handler that uses OIDC discovery to get the decode key."""
+
+    def decode_token(self, token: str, decode_key: str) -> dict:
+        alg = self._decode_complete_unverified(token)["header"]["alg"]
```

Comment:
This could fail if the payload is not well formed. Does that result in jwt.PyJWTError that is catched by the Base class?

Reply:
Results in `DecodeError` which is caught by `PyJWTError`.